### PR TITLE
Audio: Optimize sink checks in crossover_s32_default

### PR DESCRIPTION
**Pull request by ShriramShastry**
Optimize `crossover_s32_default` by pre-determining the active sinks before processing frames. This check-in reduces redundant null checks for each frame and channel, leading to performance improvements in processing audio streams with multiple sinks.

### CODE REVIEW

File: src/audio/crossover/crossover_generic.c
Comment (lyakh) [inline src/audio/crossover/crossover_generic.c:259]: indentation

File: src/audio/crossover/crossover_generic.c
Comment (lyakh) [inline src/audio/crossover/crossover_generic.c:250]: you replaced an `idx += nch` addition with an addition and a multiplication `ch + i * nch` - subject to any compiler optimisations. Depending on those optimisations this might or might not end up generating different code, but I personally don't think this would be faster. I'd keep the original addition, personally, I prefer this form:
```
for (i = 0, idx = ch; i < frames; i++, idx += nch) {
```
but that might be a matter of taste. The original code was run-time equivalent to this, don't think your version is an improvement.

File: src/audio/crossover/crossover_generic.c
Comment (lyakh) [inline src/audio/crossover/crossover_generic.c:251]: does this comment add any information?

File: src/audio/crossover/crossover_generic.c
Comment (lyakh) [inline src/audio/crossover/crossover_generic.c:256]: and now you have this calculation twice? No, please, remove this.

**Review by lyakh — DISMISSED** [review]
ok, this optimisation avoids checking inactive sinks for each frame on each channel. So yes, it should save some time when there are inactive sinks. But please restore the original index calculation

**Review by ShriramShastry — COMMENTED** [review]
Thank you for your code review feedback. I reconsidered the index increment approach and agree that incrementing the index directly within the loop construct is both stylistically clearer and potentially more efficient because it eliminates unnecessary multiplication. Regarding the comment, I recognise that it is redundant and should be removed for clarity. I'll make these changes accordingly.

File: src/audio/crossover/crossover_generic.c
Comment (lyakh) [inline src/audio/crossover/crossover_generic.c:251]: indentation - there's a spurious space in front of the comment

Comment (ShriramShastry) [reply src/audio/crossover/crossover_generic.c]: Done, Thanks.

**Review by lyakh — APPROVED** [review]

File: src/audio/crossover/crossover_generic.c
Comment (kv2019i) [inline src/audio/crossover/crossover_generic.c:229]: It would be preferred to have one logic change per git commit. E.g. these comments to existing code are not related to the the optimization, so the two should be in separate git commits for clarity. 

As for the comment, I personally like more comment heavy coding style, but in SOF we traditionally have been more conservative on the amount of comments. I'll let the crossover maintainers chime in here whether they find the comment level ok or not here. Definitely I found these helpful.

**Review by kv2019i — COMMENTED** [review]
A few minor style comments inline, but the optimization itself seems good and gets and check out of the inner loop.

File: src/audio/crossover/crossover_generic.c
Comment (cujomalainey) [reply src/audio/crossover/crossover_generic.c:229]: @johnylin76 and I are the maintainers. I second the atomic commits arguements. Please split "just adding comments" from  your actual change so its easier to understand.

File: src/audio/crossover/crossover_generic.c
Comment (cujomalainey) [inline src/audio/crossover/crossover_generic.c:239]: can you quantify the performance gain?

**Review by cujomalainey — CHANGES_REQUESTED** [review]

File: src/audio/crossover/crossover_generic.c
Comment (ShriramShastry) [reply src/audio/crossover/crossover_generic.c:228]: Done !! Please take a look.

File: src/audio/crossover/crossover_generic.c
Comment (lgirdwood) [inline src/audio/crossover/crossover_generic.c]: Best to split this, as this part is more than the commit message. I would make this a second patch that states the reason and any performance improvement data.

**Review by lgirdwood — COMMENTED** [review]
Best to split this PR in patches that each have one change. Easiest way to do this quickly is to ```git reset``` on to origin/main HEAD (this will leave all your changes as a diff and not as patches), then use ```git gui``` to individually pick lines for each patch.

File: src/audio/crossover/crossover_generic.c
Comment (cujomalainey) [inline src/audio/crossover/crossover_generic.c:238]: your commit message says "adding comments" but you are still changing code here

File: src/audio/crossover/crossover_generic.c
Comment (cujomalainey) [inline src/audio/crossover/crossover_generic.c]: all this commit does is delete the comments you added, please check your commits to make sure they make sense before pushing them.

**Review by cujomalainey — CHANGES_REQUESTED** [review]

**Review by cujomalainey — CHANGES_REQUESTED** [review]
all my previous comments are still valid, please fix them

File: src/audio/crossover/crossover_generic.c
Comment (ShriramShastry) [reply src/audio/crossover/crossover_generic.c:258]: corrected the commit message.

**Review by ShriramShastry — COMMENTED** [review]
Thank you for reviewing the change. I addressed the review comments. Please have a look.

**Review by marcinszkudlinski — APPROVED** [review]

**Review by cujomalainey — APPROVED** [review]

**Pull request merged**
